microservices/download/utils.js

```diff
@@ -111,175 +107,193 @@ async function displayError(req, res, errMsg) {
+          ) {
+
+            return returnObj;
+
+          } else if (
+            type === "linux_tar" &&
+            name.endsWith(".tar.gz") &&
+            !name.endsWith("-arm64.tar.gz")
+          ) {
+
+            return returnObj;
+
+          } else if (
+            type === "linux_rpm" &&
+            name.endsWith(".x86_64.rpm")
+          ) {
+
+            return returnObj;
+
+          } else if (
+            type === "linux_deb" &&
+            name.endsWith("_amd64.deb")
+          ) {
+
+            return returnObj;
+
           }
         }
-      }
-      return undefined;
-    };
 
-    switch(type) {
-      // Linux Binaries
-      case "linux_appimage":
-        binaryPath = findBinary(".AppImage", "end", taskGraph.data.task.artifacts[0].files);
-        break;
-      case "linux_tar":
-        binaryPath = findBinary(".tar.gz", "end", taskGraph.data.task.artifacts[0].files);
-        break;
-      case "linux_rpm":
-        binaryPath = findBinary(".rpm", "end", taskGraph.data.task.artifacts[0].files);
-        break;
-      case "linux_deb":
-        binaryPath = findBinary(".deb", "end", taskGraph.data.task.artifacts[0].files);
-        break;
-      // Windows Binaries
-      case "windows_setup":
-        binaryPath = findBinary("binaries/Pulsar Setup", "start", taskGraph.data.task.artifacts[0].files);
-        break;
-      case "windows_portable":
-        binaryPath = findBinary(".zip", "end", taskGraph.data.task.artifacts[0].files);
-        break;
-      case "windows_blockmap":
-        binaryPath = findBinary(".exe.blockmap", "end", taskGraph.data.task.artifacts[0].files);
-        break;
-      // MacOS Builds
-      case "mac_zip":
-        binaryPath = findBinary(".zip", "end", taskGraph.data.task.artifacts[0].files);
-        break;
-      case "mac_zip_blockmap":
-        binaryPath = findBinary(".zip.blockmap", "end", taskGraph.data.task.artifacts[0].files);
-        break;
-      case "mac_dmg":
-        binaryPath = findBinary(".dmg", "end", taskGraph.data.task.artifacts[0].files);
-        break;
-      case "mac_dmg_blockmap":
-        binaryPath = findBinary(".dmb.blockmap", "end", taskGraph.data.task.artifacts[0].files);
-        break;
-      default:
-        binaryPath = undefined;
-        break;
+      }
     }
 
-    if (binaryPath === undefined) {
-      return {
-        ok: false,
-        code: 503,
-        msg: "Invalid TYPE Download Parameters Provided"
-      };
-    }
+    // If we get to this point it means the above loop didn't return.
+    // Meaning we couldn't find a single valid asset among any versions
+    // So we will return an error
 
-    // Now that we have the binary, it's time to return a redirect.
     return {
-      ok: true,
-      content: `https://api.cirrus-ci.com/v1/artifact/task/${taskid}/binary/${binaryPath}`
+      ok: false,
+      code: 404,
+      msg: `Unable to find any assets matching the provided parameters: os=${os};type=${type}`
     };
```

Comment:
NO ACTION NEEDED, JUST COMMENTING 👍 

This "asset not found" handling _does_ work in my testing. Some of the blockmap asset types aren't available, like the intel macOS and the Windows ones, and if you request one of those, the microservice legitimately can't find them, and the error message is returned as it should be in that situation. 👍 

Not the microservice's fault those assets genuinely aren't available, so 404'ing is the right move. 👍  And it does work as such in my testing. 👍

Reply:
Part of the reason I originally didn't upload them, made the 404 easy to test, so happy you were able to confirm